File: daily_problems/2026/07/0709/personal_submission/cf106054m_yawn_sean.py
# Submission link: https://codeforces.com/gym/106054/submission/381676346
def main():
    n, m, k = MII()
    path = [[] for _ in range(n)]
    
    for _ in range(m):
        u, v = GMI()
        path[u].append(v)
        path[v].append(u)
    
    dis = [-1] * n
    dis[0] = 0
    
    que = [0]
    
    for u in que:
        for v in path[u]:
            if dis[v] == -1:
                dis[v] = dis[u] + 1
                que.append(v)
    
    if dis[1] == -1:
        print(0)
        exit()
    
    mod = 998244353
    f = Factorial(10000, mod)
    
    dp = [[0] * n for _ in range(k)]
    
    dp[0][0] = 1
    
    for i in range(k):
        for u in que:
            if dp[i][u] == 0:
                continue
            
            for v in path[u]:
                ni = dis[u] + i + 1 - dis[v]
                if ni < k:
                    dp[ni][v] += dp[i][u]
                    dp[ni][v] %= mod
    
    ans = 0
    days = (dis[1] - 1) // k + 1
    
    for i in range(k):
        if dis[1] + i <= days * k:
            res = 0
            
            for cnt in range(days + 1):
                resid = dis[1] + i - cnt * k
                if cnt % 2 == 0:
                    res += f.combi(resid - 1, days - 1) * f.combi(days, cnt) % mod
                else:
                    res += mod - f.combi(resid - 1, days - 1) * f.combi(days, cnt) % mod
                res %= mod
            
            ans += res * dp[i][1] % mod
            ans %= mod
    
    print(ans)

File: daily_problems/2026/07/0709/personal_submission/cf106054m_yawn_sean.cpp
#include <bits/stdc++.h>
// #include "atcoder/all"
// #pragma GCC optimize("Ofast")
// #pragma GCC optimize("unroll-loops")
// #pragma GCC target("sse,sse2,sse3,ssse3,sse4,popcnt,abm,mmx,avx2,tune=native")

using namespace std;
auto rng = mt19937(random_device()());
auto rngl = mt19937_64(random_device()());

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(0);
	cout.tie(0);

	int n, m, k;
	cin >> n >> m >> k;

	vector<vector<int>> path(n);

	while (m --) {
		int u, v;
		cin >> u >> v;
		u --, v --;
		path[u].emplace_back(v);
		path[v].emplace_back(u);
	}

	vector<int> dis(n, -1), que = {0};
	dis[0] = 0;

	for (int i = 0; i < que.size(); i ++) {
		int u = que[i];
		for (auto &v: path[u]) {
			if (dis[v] == -1) {
				dis[v] = dis[u] + 1;
				que.emplace_back(v);
			}
		}
	}

	if (dis[1] == -1) cout << 0;
	else {
		int mod = 998244353, bound = n + k;

		vector<vector<int>> combs(bound, vector<int>(bound, 0));

		for (int i = 0; i < bound; i ++) {
			combs[i][0] = 1;
			combs[i][i] = 1;

			for (int j = 1; j < i; j ++) {
				combs[i][j] = (combs[i - 1][j] + combs[i - 1][j - 1]) % mod;
			}
		}

		auto comb = [&] (int x, int y) -> int {
			if (x < 0 || y < 0 || y > x) return 0;
			return combs[x][y];
		};

		vector<vector<int>> dp(k, vector<int>(n, 0));
		dp[0][0] = 1;

		for (int i = 0; i < k; i ++) {
			for (auto &u: que) {
				if (dp[i][u] == 0) continue;

				for (auto &v: path[u]) {
					int ni = dis[u] + i + 1 - dis[v];
					if (ni < k) {
						dp[ni][v] += dp[i][u];
						dp[ni][v] %= mod;
					}
				}
			}
		}

		int ans = 0, days = (dis[1] - 1) / k + 1;

		for (int i = 0; i < k; i ++) {
			if (dis[1] + i <= days * k) {
				int res = 0;

				for (int cnt = 0; cnt <= days; cnt ++) {
					int resid = dis[1] + i - cnt * k;
					if (cnt % 2 == 0)
						res += 1ll * comb(resid - 1, days - 1) * comb(days, cnt) % mod;
					else
						res += mod - 1ll * comb(resid - 1, days - 1) * comb(days, cnt) % mod;
					res %= mod;
				}

				ans += 1ll * res * dp[i][1] % mod;
				ans %= mod;
			}
		}

		cout << ans;
	}

	return 0;
}